### ISSUE TITLE
ValueError: invalid __array_struct__ when using js arrays of arrays and numpy

### ISSUE THREAD

**Issue by jstafford** [iodide-project/pyodide#316]
When using a matrix (array of array of numbers) in javascript and trying to convert that to a numpy array, it fails with the error `ValueError: invalid __array_struct__`

To reproduce:
JavaScript:
```
window.A = [[1,2,3],[4,5,6]];
```
Python:
```
import numpy
from js import A
m = numpy.array(A)
```

**Comment by jstafford** [comment]
Interestingly, other libraries besides numpy do seem to be able to work with JavaScript matrices. For example, statistics does work with a JavaScript matrix.
JavaScript:
```
window.A = [[1,2,3],[4,5,6]];
```
Python:
```
import statistics
from js import A
statistics.mean(A[0])
```

**Comment by jstafford** [comment]
After some more digging around in the code, it appears `__array_struct__` is a special `numpy` thing.
Looking in the code for the numpy package at `packages/numpy`, I see that it is always accessed using the function `PyArray_LookupSpecial_OnInstance`. When I look at that function, it has the interesting comment: 
```
 * Implements incorrect special method lookup rules, that break the python
 * convention, and looks on the instance, not the type.
 *
 * Kept for backwards compatibility. In future, we should deprecate this.
```
This seems like it could be the root of the problem, but it is unclear to me what change should be made. Patching the implementation of `PyArray_LookupSpecial_OnInstance` to be that of `PyArray_LookupSpecial` might make it work for JavaScript objects, but could break it for native python objects. It's very hard to follow how this code works.

**Comment by jstafford** [comment]
More searching the code reveals that the error must be produced at line 2241 of `packages/numpy/build/numpy-1.15.1/numpy/core/src/multiarray/ctors.c`. This means that the call to `PyArray_LookupSpecial_OnInstance` at line 2201 of `ctors.c` must be returning something which is not `NULL`, but is also not a valid `__array_struct__`. Still unclear to me how/why this is happening.

**Comment by mdboom** [comment]
I can't reproduce this as of Pyodide 0.8.2 -- it's possible the changes to object coercion helped.

However, even given that, I wouldn't expect Numpy to be able to automatically convert this given how Javascript Arrays and Objects are kinda-sorta the same thing.  I think the best we could expect is to require:

```
np.array([list(x) for x in A])
```

(and maybe provide as `aslist` helper function to handle nested lists).
